Userland/xargs.cpp

```diff
@@ -0,0 +1,276 @@
+        ++index;
+    }
+
+    // Append a '{}' at the end if no arguments have replacements.
+    if (!some_argument_has_replacement) {
+        Vector<StringView> parts;
+        parts.append("");
+        parts.append("");
+        argument_parts.append(move(parts));
+    }
+
+    FILE* fp = stdin;
+    bool is_stdin = true;
+
+    if (StringView { "-" } != file_to_read) {
+        // A file was specified, try to open it.
+        fp = fopen(file_to_read, "r");
+        if (!fp) {
+            perror("fopen");
+            return 1;
+        }
+        is_stdin = false;
+    }
+
+    char* line = nullptr;
+    size_t buffer_size = 0;
+
+    StringBuilder builder;
+    Vector<char*> child_argv;
+
+    int devnull_fd = 0;
+
+    if (is_stdin) {
+        devnull_fd = open("/dev/null", O_RDONLY | O_CLOEXEC);
+        if (devnull_fd < 0) {
+            perror("open");
+            return 1;
+        }
+    }
+
+    bool fail = false;
+    int lines_used_for_this_command = 0;
+    char* last_line = nullptr;
+    size_t last_line_size = 0;
+    size_t total_command_length = 0;
+
+    for (;;) {
+        line = nullptr;
+        buffer_size = 0;
+
+        ssize_t line_size;
+
+        if (last_line) {
+            line_size = last_line_size;
+            line = last_line;
+            last_line = nullptr;
+        } else {
+            line_size = getdelim(&line, &buffer_size, entry_separator, fp);
+        }
+
+        if (line_size < 0) {
+            if (errno != 0) {
+                perror("getdelim");
+                fail = true;
+            }
+            break;
+        }
+
+        last_line = line;
+        last_line_size = line_size;
+
+        if (line_size == 0) {
+            free(line);
+            last_line = nullptr;
+            continue;
+        }
+
+        ++lines_used_for_this_command;
+
+        if (line_size > 0) {
+            while (line[line_size - 1] == '\n')
+                --line_size;
+
+            if (!line_size) {
+                last_line = nullptr;
+                continue;
+            }
+
+            StringView line_view { line, (size_t)line_size };
+
+            dbg() << "Getline '" << line_view << "'";
+
+            size_t index = 0;
+            if (lines_used_for_this_command == 1) {
+                child_argv.ensure_capacity(argument_parts.size());
+                for (auto& parts : argument_parts) {
+                    builder.clear();
+                    builder.join(line_view, parts);
+                    auto string = builder.build();
+                    total_command_length += string.length();
+                    child_argv.unchecked_append(strdup(string.characters()));
+                    ++index;
+                }
+            } else {
+                if ((max_lines_for_one_command > 0 && lines_used_for_this_command > max_lines_for_one_command) || total_command_length + line_size + 1 >= max_bytes) {
+                    run_command(child_argv, verbose, is_stdin, devnull_fd);
```

Comment:
You're running the command as is after you've established that `lines_used_for_this_command > max_lines_for_one_command`. You're not actually using the line for this command, are you? Then don't increment `lines_used_for_this_command` until you do decide that you are definetely going to use it.

Reply:
it's reset to zero below, then set to 1 on the next iteration for the next command.
it's easier to write out like this, without `lines_used_for_this_command + 1`.

Reply:
I guess its name is a bit misleading, it would be `lines_that_would_be_used_for_this_command`, but that's so damn long

Reply:
And yes, it sounds like the code would benefit a lot from untangling "read more lines until <condition>" from "construct child argv from the lines being read". The former part doesn't have to know about `child_argv`, `lines_used_for_this_command`, etc — just read lines and invoke a callback (perhaps the callback would either accept or "unget" the line). The latter part doesn't have to know about `getdelim`, separators, `last_line`, `line_size`, all that crap — it just sees if the new line fits and if so, append it to the argv, otherwise reject it and `run_command`.